src/katgpucbf/fgpu/kernels/postproc.mako

```diff
@@ -42,20 +42,42 @@ DEVICE_FN float2 csub(float2 a, float2 b)
     return make_float2(a.x - b.x, a.y - b.y);
 }
 
-DEVICE_FN char quant(float value)
+// Quantise, and return saturation flag
+DEVICE_FN bool quant(float value, signed char *out)
 {
-    int out;
+    bool saturated = false;
+    if (value < -127.0f)
+    {
+        value = -127.0f;
+        saturated = true;
+    }
+    else if (value > 127.0f)
+    {
+        value = 127.0f;
+        saturated = true;
+    }
+
+    int q;
+    // Convert to s8, round to nearest integer, and saturate
+    // (saturation is redundant due to the test above, but automatic
+    // for float-to-int conversions on CUDA hardware).
 #ifdef __OPENCL_VERSION__
-    out = convert_char_sat_rte(value);
+    q = convert_char_sat_rte(value);
 #else
-    // Convert to s8, round to nearest integer, and saturate
-    asm("cvt.rni.sat.s8.f32 %0, %1;" : "=r" (out) : "f"(value));
+    asm("cvt.rni.sat.s8.f32 %0, %1;" : "=r" (q) : "f"(value));
 #endif
-    // Clamp -128 to -127 to give symmetrical output
-    out = (out == -128) ? -127 : out;
-    return out; // Returning an int as a char can lose values, but since we have
-                // just made sure it's s8 in the previous line, it's guaranteed
-                // not to.
+    // Returning an int as a char can lose values, but since we have
+    // just made sure it's s8 in the previous line, it's guaranteed
+    // not to.
+    *out = q;
+    return saturated;
+}
+
+// Quantise, and return saturation flag
+DEVICE_FN bool quant_complex(float2 value, signed char *out_x, signed char *out_y)
+{
+    // Note: | not || to avoid short-circuiting
```

Comment:
> Let me make sure I'm understanding correctly: if `quant(value.x, out_x)` was truthy then `quant(value.y, out_y)` wouldn't even be bothered with?

That's what would happen if you used `||`, but it would be bad because we want the side effect of `quant` (writing to the output). Using `|` ensures both calls to `quant` always happen.

> Interesting approach. I'd probably have done something like
> 
> ```c++
> a = quant(x)
> b = quant(y)
> return a || b
> ```
> 
> but I suppose your approach works as well.
> 
> I'm a bit hazy about how things work in constrained environments like GPU kernels but I'd imagine that fewer variables in the stack is probably better. Even if they're only short-lived? Or am I imagining things?

I did it my way purely for compactness of the source. I expect the compiler will almost certainly produce identical code. If you think your approach is clearer I'm happy to change it (this isn't code golf), although I'd still want to put a warning in to dissuade anyone from attempting to compact it to `quant(value.x, out_x) || quant(value.y, out_y)`.

Reply:
Understood about the side-effects being necessary. The approach I wrote down is just how I'd have solved the problem, not to suggest it was a better solution. So I'm happy to leave it as it is, I think your comment is sufficient for the casual reader to understand.
@amishatishpatel, any opinions?